File: dynamic_programming/linear_dp/single_chain/0300_longest_increasing_subsequence.js
/*
Given an integer array nums, return the length of the longest strictly 
increasing subsequence.

A subsequence is an array that can be derived from another array by deleting 
some or no elements without changing the order of the remaining elements.

1 <= nums.length <= 2500
-10^4 <= nums[i] <= 10^4
*/

// https://leetcode.com/problems/longest-increasing-subsequence/

// brute force -- TLE
var lengthOfLIS = function (nums) {
  var findLISStartAtIndex = function (nums, index) {
    if (index === nums.length - 1) return 1;
    let maxEndingHere = 1;
    for (let i = index + 1; i < nums.length; i++) {
      if (nums[index] < nums[i]) {
        maxEndingHere = Math.max(
          maxEndingHere,
          1 + findLISStartAtIndex(nums, i)
        );
      }
    }
    return maxEndingHere;
  };
  let ans = 1;
  for (let i = 0; i < nums.length; i++) {
    ans = Math.max(ans, findLISStartAtIndex(nums, i));
  }
  return ans;
};

// memoization
var lengthOfLIS = function (nums) {
  let memo = new Map();
  var findLISStartAtIndex = function (nums, index) {
    if (index === nums.length - 1) return 1;
    if (memo.has(index)) return memo.get(index);
    let max = 1;
    for (let i = index + 1; i < nums.length; i++) {
      if (nums[index] < nums[i]) {
        max = Math.max(max, 1 + findLISStartAtIndex(nums, i));
      }
    }
    memo.set(index, max);
    return max;
  };
  let ans = 1;
  for (let i = 0; i < nums.length; i++) {
    ans = Math.max(ans, findLISStartAtIndex(nums, i));
  }
  return ans;
};

// optimized dynamic programming
// let dp[i] be the length of the LIS that ends with nums[i]
// for every j that 0 <= j < i; if nums[j] < nums[i], dp[i] = max(dp[i], dp[j]+1)
// time complexity: O(n^2)
// space complexity: O(n)
var lengthOfLIS = function (nums) {
  const dp = new Array(nums.length).fill(1);
  for (let i = 1; i < nums.length; i++) {
    for (let j = 0; j < i; j++) {
      if (nums[j] < nums[i]) {
        dp[i] = Math.max(dp[i], dp[j] + 1);
      }
    }
  }
  return Math.max(...dp);
};

File: dynamic_programming/linear_dp/single_chain/0300_longest_increasing_subsequence.py
# Given an integer array nums, return the length of the longest strictly 
# increasing subsequence.

# A subsequence is an array that can be derived from another array by deleting 
# some or no elements without changing the order of the remaining elements.

# 1 <= nums.length <= 2500
# -10^4 <= nums[i] <= 10^4

# https://leetcode.com/problems/longest-increasing-subsequence/

# brute force 
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        @cache
        def LIS_start(start_idx):
            if start_idx == len(nums): return 1
            ans, l = 1, len(nums)
            for i in range(start_idx + 1, l):
                if nums[start_idx] < nums[i]:
                    ans = max(ans, 1 + LIS_start(i))
            return ans

        ans, l = 1, len(nums)
        for i in range(l):
            ans = max(ans, LIS_start(i))
        return ans

# memoization
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        table = dict()
        def LIS_start(start_idx):
            # check memo first
            if start_idx in table:
                return table[start_idx]
            # base case
            if start_idx == len(nums): return 1
            # recursive case
            ans, l = 1, len(nums)
            for i in range(start_idx + 1, l):
                if nums[start_idx] < nums[i]:
                    ans = max(ans, 1 + LIS_start(i))
            table[start_idx] = ans
            return ans

        ans, l = 1, len(nums)
        for i in range(l):
            ans = max(ans, LIS_start(i))
        return ans
    
# dynamic programming O(n^2)
# dp[i] denotes the length of LIS that ends with nums[i]
# for every j that 0 <= j < i and nums[j] < nums[i]: dp[i] = max(dp[i], dp[j] + 1)
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        l = len(nums)
        dp = [1] * (l + 1)
        for i in range(1, l):
            for j in range(0, i):
                if nums[j] < nums[i]:
                    dp[i] = max(dp[i], dp[j] + 1)
        return max(dp)
    
# greedy + binary search O(nlogn)
from bisect import bisect_left
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        if not nums:
            return 0
        LIS = []
        for num in nums:
            # Use binary search to find the index of the first element in LIS that is >= num
            idx = bisect_left(LIS, num)
            # If num is larger than any element in LIS, append it
            if idx == len(LIS):
                LIS.append(num)
            else:
                # Replace the existing element to maintain the smallest possible value
                LIS[idx] = num
        # The length of LIS will be the length of the Longest Increasing Subsequence
        return len(LIS)